src/topology/builder.rs

```diff
@@ -275,16 +273,10 @@ pub async fn build_pieces(
     }
 }
 
-fn filter_event_type<S>(
-    stream: S,
-    data_type: DataType,
-) -> Box<dyn Stream01<Item = Event, Error = ()> + Send>
-where
-    S: Stream01<Item = Event, Error = ()> + Send + 'static,
-{
+fn filter_event_type(event: &Event, data_type: DataType) -> bool {
     match data_type {
-        DataType::Any => Box::new(stream), // it's possible to not call any comparing function if any type is supported
-        DataType::Log => Box::new(stream.filter(|event| matches!(event, Event::Log(_)))),
-        DataType::Metric => Box::new(stream.filter(|event| matches!(event, Event::Metric(_)))),
+        DataType::Any => true,
+        DataType::Log => matches!(event, Event::Log(_)),
+        DataType::Metric => matches!(event, Event::Metric(_)),
     }
 }
```

Comment:
I would tend to think this a step backwards, for performance at least. Can this still be done with a stream adapter like before for all the places it's used above?

Reply:
It looked very suspicious to me too - that the code before the change was duplicated like that, but I don't get the idea behind it. Could you please elaborate more on where will the performance difference come from?

Reply:
Thinking about it a bit more, it probably doesn't make a big difference on modern CPUs, and might actually be a win. Before, each pipeline would be instantiated with a fixed stream filter function (single conditional branch on the event type) but will likely require an indirect jump to get to that function. With this change, each pipeline is instantiated with a generic stream filter function (two conditional branches). Mispredicted indirect jumps are likely more expensive than mispredicted branches. This is pretty down in the weeds of micro-benchmarking, though.

Reply:
Well, regarding performance, with the old there would have to be two dyn boxing which this replaces for an integer switch. Two data memory jumps + two code jumps vs one local integer switch.

Reply:
Nit: the replacement code has two switches: first on `data_type` and then on the `matches!` inside the match.

Reply:
That's true I just compared the diff.